### PR TITLE
Issue 1761: Differing segment count/split/merges metrics for same stream

### LINKED ISSUE #1761: Multiple controllers send differing segment count/splits/merges metrics on the same stream
**Problem description**

On clusters with two Pravega controller instances, if two controllers are reporting `segments_count`, `segment_splits`, and `segment_merges` metrics on the same stream, the numbers being reported differ between the controllers.

One of two controllers would report stale metrics as of the moment when the stream was created (initial number of segments for `segments_count`, and zeros for `segment_splits` and `segment_merges`).
The second controller reports the actual number of active segments in `segments_count`, but much bigger numbers than the actual metrics for `segment_splits` and `segment_merges` would have to be.

###### Example:
One controller is reporting 
```
53e392ef-27ad-4b8b-be0b-198a14b7b7b1-0004.pravega_controller__b9051a32-5ff8-4083-a343-702f4ae1a34a.66cd021c-23da-46eb-88a3-d64166d5c1ab.pravega.controller.segments_count.hulk.smallScale.Gauge:1|g
```
The other controller is reporting
```
53e392ef-27ad-4b8b-be0b-198a14b7b7b1-0004.pravega_controller__9ea16bb9-05b6-424e-9066-e799a51b9134.5810a308-b5f6-4adf-ac65-b6a680cb7fbb.pravega.controller.segments_count.hulk.hulk.smallScale.Gauge:15|g
```
*(the correct number of active segments is 15, as confirmed by `scaling-events` API)*

###### Another example:
One controller is reporting 
```
c41e8d11-b058-4bf1-883a-829185d8134d-0004.pravega_controller__b5eed754-a70a-4200-8299-1c570b57902b.67f3abd2-4365-45bf-9bc9-3b9184b5330c.pravega.controller.segment_splits.hulk.smallScale.Counter:0|g
```
The other controller is reporting
```
c41e8d11-b058-4bf1-883a-829185d8134d-0004.pravega_controller__2a23b6bb-cca4-4543-a432-991d8fa62b6d.48d93a29-27f7-4257-9711-c4c5f1fa1c09.pravega.controller.segment_splits.hulk.smallScale.Counter:2453|g
``` 
*(the correct number would be below 400, as the total number of ever created segments is less than 400 for the stream)*

If only one of the two controllers is reporting the metrics for a given stream, then all the three metrics are reported correctly.

**Problem location**
Pravega controller - multiple instances

**Suggestions for an improvement**
Multiple controllers should report identical metrics;
they should report correct counters for `segment_splits` and `segment_merges`.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
This PR removes the initialization of `segment_split` and `segment_merges` metrics upon a Stream creation in the Controller. Now we have `initial_segments_count` and the existing `segments_count` to avoid conflicts among Controller instances reporting the segments of a Stream. Fixed reporting of `segment_split` and `segment_merges`, as they should be gauges (not counters).

**Purpose of the change**  
Fixes #1761.

**What the code does**  
The original problem was that multiple Controller instances could be reporting concurrently count/split/merges metrics for same Stream. This leads to bogus metrics values. The reason for this to happen is that, upon a Stream creation, the Controller was initializing `segment_split`, `segment_merges` and `segment_count` metrics. This yields that the instance handling the Stream creation starts reporting these metrics. Then, if another Controller instance handles an operation on the same Stream that involves reporting these metrics (e.g., scale), then the conflict appears.

To solve this problem, this PR contains the following changes:
1.- Removed from `createStream` the initialization of `SEGMENTS_SPLITS` and `SEGMENTS_MERGES`.
2.- Now, we have 2 metrics to report the segments of a Stream; one called `initial_segments_count` and the existing `segments_count`. In `initial_segments_count`, the Controller handling the Stream creation will publish the initial number of segments for a Stream. Then, the Controller handling scale events for that Stream will publish the active segments values in `segments_count` metric, as regular. This avoids conflicts across Controllers and we use metrics in the way they are supposed to be used (values published by an individual process, no coordination).
3.- As an additional problem, `SEGMENTS_SPLITS` and `SEGMENTS_MERGES` were being reported as counters. However, they should be gauges, as the values that is being published are already "totals" provided by the Controller.
4.- Moreover, to avoid conflicts on Controller restarts, we delete all the metrics and close the stats provider when the Controller shuts down. For this reason, this PR moves the initialization of metrics from `Main.java` to `ControllerServiceStarter.java`.

**How to verify it**  
It has been tested (locally and in a distributed deployment) that segment count/split/merges are being reported by a single Controller instance. All tests should be passing as before.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 1]
```diff
@@ -57,10 +76,11 @@
...
     public void createStream(String scope, String streamName, int minNumSegments, Duration latency) {
         DYNAMIC_LOGGER.incCounterValue(CREATE_STREAM, 1);
-        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(OPEN_TRANSACTIONS, scope, streamName), 0);
+        // We need to initialize the number of initial segments, but prevent this process to continue reporting the
+        // segment count for this Stream. This will avoid multiple controllers to report conflicting values.
         DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName), minNumSegments);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, streamName), 0);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_MERGES, scope, streamName), 0);
+        executor.schedule(() -> DYNAMIC_LOGGER.freezeGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName)),
+                statsOutputFrequencySeconds + 1, TimeUnit.SECONDS);
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: Why "+1" ?
Also, what does freeGaugeValue do?

Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: `freezeGaugeValue` removes the metric from the metrics cache, so it is not reported anymore. By delaying the execution `freezeGaugeValue` one reporting period, we ensure that we report the initial number of segments for this Stream and then the Controller will stop reporting it. Regarding the `+ 1`, it was just to be sure that the deletion occurs at least after a full `statsOutputFrequencySeconds` period completes, but maybe I can delete it for clarity.

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 2]
```diff
@@ -38,14 +44,27 @@
...
 
+    private final int statsOutputFrequencySeconds;
+
     private final OpStatsLogger createStreamLatency = STATS_LOGGER.createStats(CREATE_STREAM_LATENCY);
     private final OpStatsLogger deleteStreamLatency = STATS_LOGGER.createStats(DELETE_STREAM_LATENCY);
     private final OpStatsLogger sealStreamLatency = STATS_LOGGER.createStats(SEAL_STREAM_LATENCY);
     private final OpStatsLogger updateStreamLatency = STATS_LOGGER.createStats(UPDATE_STREAM_LATENCY);
     private final OpStatsLogger truncateStreamLatency = STATS_LOGGER.createStats(TRUNCATE_STREAM_LATENCY);
 
+    private ScheduledExecutorService executor = Executors.newScheduledThreadPool(2);
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: I don't see a reason to introduce another thread pool. More fundamentally, I don't see a reason for having a reporting period. The problem I have observed is that values reported are not monotonically increasing, which I'm not sure why it is happening.

Comment (andreipaduroiu) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: In addition to what @fpj said:
1. This thread pool is not created using `ExecutorServiceHelpers`. The factory methods in there take care of a number of other things for you.
2. This is not a final variable

I recommend you reuse an existing threadpool and pass it via the constructor. That way we don't have idle threads sitting around for most of the time.

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 3]
```diff
@@ -57,10 +76,11 @@
      */
     public void createStream(String scope, String streamName, int minNumSegments, Duration latency) {
         DYNAMIC_LOGGER.incCounterValue(CREATE_STREAM, 1);
-        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(OPEN_TRANSACTIONS, scope, streamName), 0);
+        // We need to initialize the number of initial segments, but prevent this process to continue reporting the
+        // segment count for this Stream. This will avoid multiple controllers to report conflicting values.
         DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName), minNumSegments);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, streamName), 0);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_MERGES, scope, streamName), 0);
+        executor.schedule(() -> DYNAMIC_LOGGER.freezeGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName)),
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: You run this every few seconds. If you don't create the thread pool using `ExecutorServiceHelpers`, this task will continue running after you invoke `shutdown`.

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 4]
```diff
@@ -57,10 +78,13 @@
...
     public void createStream(String scope, String streamName, int minNumSegments, Duration latency) {
         DYNAMIC_LOGGER.incCounterValue(CREATE_STREAM, 1);
-        DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(OPEN_TRANSACTIONS, scope, streamName), 0);
+        // We need to initialize the number of initial segments, but prevent this process to continue reporting the
+        // segment count for this Stream. This will avoid multiple controllers to report conflicting values.
         DYNAMIC_LOGGER.reportGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName), minNumSegments);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, streamName), 0);
-        DYNAMIC_LOGGER.incCounterValue(nameFromStream(SEGMENTS_MERGES, scope, streamName), 0);
+        if (executor != null) {
+            executor.schedule(() -> DYNAMIC_LOGGER.freezeGaugeValue(nameFromStream(SEGMENTS_COUNT, scope, streamName)),
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: What happens if this stream never splits or merges (e.g., fixed policy)?

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java [hunk 5]
```diff
@@ -210,10 +234,10 @@ public static void reportActiveSegments(String scope, String streamName, int num
      * @param merges        Number of segment merges in the scale operation.
      */
     public static void reportSegmentSplitsAndMerges(String scope, String streamName, long splits, long merges) {
-        DYNAMIC_LOGGER.updateCounterValue(globalMetricName(SEGMENTS_SPLITS), splits);
-        DYNAMIC_LOGGER.updateCounterValue(nameFromStream(SEGMENTS_SPLITS, scope, streamName), splits);
-        DYNAMIC_LOGGER.updateCounterValue(globalMetricName(SEGMENTS_MERGES), merges);
-        DYNAMIC_LOGGER.updateCounterValue(nameFromStream(SEGMENTS_MERGES, scope, streamName), merges);
+        DYNAMIC_LOGGER.reportGaugeValue(globalMetricName(SEGMENTS_SPLITS), splits);
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: Can we add one or more test cases to validate that now the behavior is the one expected?

File: controller/src/main/java/io/pravega/controller/server/ControllerService.java [hunk 6]
```diff
@@ -78,6 +79,7 @@
     private final StreamMetrics streamMetrics;
     private final TransactionMetrics transactionMetrics;
 
+    @VisibleForTesting
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/ControllerService.java]: This is not visible for testing as it was public already.

File: controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java (hunk 4 above)
Comment (shiveshr) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: In that case this value is not reported after the first time.

Comment (fpj) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: but what happens to the service relying on this metrics. Does it get stored permanently on something like InfluxDB?

Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/metrics/StreamMetrics.java]: Yes, the StatsD reporter can store that metric to a persistent time-series database, such as InfluxDB. As @shiveshr says, the behavioral difference is that before this PR, the Controller handling the `createStream` would be reporting the "initial number of segments" for a given Stream for a period of time (i.e., `dynamicCacheEvictionDurationMs`). With this PR, the Controller handling the `createStream` operation will only report the "initial number of segments" for a Stream only once. With this, we achieve i) to report that information, ii) and do not make that Controller to "smash" the (updated) information being reported by another Controller that handles a `scaleStream` request for that Stream.

File: controller/src/main/java/io/pravega/controller/server/ControllerService.java (hunk 6 above)
Comment (RaulGracia) [reply controller/src/main/java/io/pravega/controller/server/ControllerService.java]: I introduced this constructor in #3117 to avoid all the existing test classes instantiating `ControllerService` to be changed just because of the two new fields (`streamMetrics`, `transactionMetrics`), which are in fact not relevant for most test classes. As this constructor is used only in testing environment, I thought that this annotation may make that more explicit. But I can remove it if you believe that it is not worth it.

**Comment by RaulGracia** [comment]
After some discussions, I provide an alternative approach to solve the issue (originally proposed by @shiveshr): create one metric for the initial segment number of a Stream (`initial_count`), and keep the `segment_count` metric as it is. This will resolve the conflict among Controllers without introducing exceptions in the handling of metrics. According to @jkhalack, this may be a feasible alternative approach to show the number of segments of a Stream in a dashboard.

**Comment by RaulGracia** [comment]
Currently, we have 2 alternative approaches to solve this problem. From the Pravega side, both are simple to implement. However, this may not be the case from the dashboard perspective.

1) _Two metrics approach_: As mentioned before, we currently create one metric for the initial segment number of a Stream (`initial_count`), and keep the `segment_count` metric as it is. This will resolve the conflict among Controllers without introducing exceptions in the handling of metrics.

From the dashboard side, this may require to build the the segment count plot showing two metrics (`initial_count`, `segment_count`) to see all the information related to a Stream's segments.

2) _Publish epoch + segment count in a single metric_: A recent proposal from @fpj is to use only one metric, instead of two. To resolve the original problem of two Controllers publishing conflicting `segment_count` values for a Stream, we need to include the `epoch` in the value published. Thus, the correct number of active segments corresponds the value with highest epoch. 

I tried this approach. As an example, below the Controller publishes `segment_count` values for `_abortStream` like (`{epoch}.{active_segments}`):
```
t    value
1546542868    0.2
1546542928    0.2
1546542988    0.2
1546543048    0.2
``` 
In this case, there is only one Controller reporting values for `_abortStream` that is in `epoch=0`, and currently has `2 segments`. One can imagine that in the case of having one Controller publishing `0.2` and another one publishing `1.1`, there is no conflict as the highest epoch is the one that represent the current number of active segments (e.g., in this case, `1`).

The complexity resides on the dashboard side, as we need to i) select the max value, and ii) use only the "second member" or "decimal part" of the published values to plot in Grafana, as is the piece of information that indicates the number of segments. At the high level, we could isolate the number of segments in the value with the highest epoch by querying InfluxDB with something like: `select max(segment_count) - floor(max(segment_count)) from segment_count ...`. This would isolate the decimal part of the max value in the table, which corresponds to the current number of segments. 

However, I would need your confirmation @EronWright @jkhalack that this may be enough to present a meaningful plot to users.

**Comment by jkhalack** [comment]
Reporting the `segment_count` as `{epoch}.{active_segments}` might be not a good idea, since there is no way to distinguish whether `0.2` reports `2`, `20`, or `200` segments.

**Comment by RaulGracia** [comment]
Thanks @jkhalack. Does it mean that the current approach with 2 metrics is acceptable from a Dashboard perspective or you have any additional suggestion?

**Comment by RaulGracia** [comment]
@jkhalack @EronWright @fpj Let me summarize the situation of this PR, so we can make a decision on how to proceed. For context, the implementation of this PR has evolved over 4 approaches:
1) _Automatically evicting `segment_count` metric from Controller handling Stream creation_: In this first approach, the Controller in charge of creating the Stream was instructed to publish the initial segment count only once in the `segment_count` metric. With this, we could have the initial segments of a Stream published in `segment_count`, and avoid the Controller creating the Stream to publish conflicting values with the Controller in charge of an hypothetical scale event for that Stream. This was done by automatically evicting the `segment_count` metric from the Controller's metrics cache that created the Stream after `dynamicCacheEvictionDurationMs`. However, with this approach it was expected but not guaranteed that the number of initial segments was actually published. Even worse, doing this was an exception in the way Pravega publishes metrics, which represents additional complexity and maintenance effort. 

2) _Conflating epoch + active segments in `segment_count`_: A second approach to solve the problem of two Controllers publishing conflicting `segments_count` values was to add the `epoch` as a discriminator. This was done by publishing values in the form `epoch.active_segments`. That is, even if two Controllers published that metric for a given Stream, we only need to plot the value with the highest `epoch`. While this was easy to implement in the Pravega side, the limitations in the dashboard side makes it hard to parse and correctly plot the active segments for those `epoch.active_segments` values.

3) _REST API for `segments_count` metric_: Another proposal has been to create the `segments_count` dashboard for a Stream from the REST API in the Controller. However, in this case the limitations of the technologies used in the dashboard-side prevents us from executing REST calls. 

4) _Two metrics approach_: This is the current implementation of this PR. We have 2 metrics, one called `initial_segments_count` and the existing `segments_count`. In `initial_segments_count`, the Controller handling the Stream creation will publish the initial number of segments for an Stream. Then, the Controller handling scale events for that Stream will publish the active segments values in `segments_count` metric, as it happens today. This avoids conflicts across Controllers and we use metrics in the way they are supposed to be used (values published by an individual process, no coordination).

Given that, I would like to hear your feedback in this regard. Note that this PR also fixes other problems, such as the fact that `segment_splits` and `segment_merges` are incorrectly being published as counters, instead of as gauges. Please, let me know if the current approach is sufficient for now (it may be improved in the future) or if there is any concrete alternative that would also fit our needs.

**Comment by andreipaduroiu** [comment]
@RaulGracia @fpj  What is status of this? I haven't seen any activity in almost 2 months.

**Comment by RaulGracia** [comment]
@andreipaduroiu thanks for asking. This PR is stuck waiting for reaching a consensus on the solution to implement the reporting of `segment_count` metric (see my previous comment summarizing the issue). Hope we can unblock it soon; once we have a decision, I will resolve the conflicts.

**Comment by jkhalack** [comment]
This has not been merged yet?

**Comment by RaulGracia** [comment]
@jkhalack I'm applying the changes to `master` in a new branch, given that there were some conflicts. I will execute a system test build to verify that the changes work and, in parallel, could you verify that the metrics are being reported properly?

**Comment by jkhalack** [comment]
@RaulGracia Do you plan to update the Micrometer metrics logic too?

There is another PR (#3504) to reorganize (introduce) the metrics tags, and I've noticed that the splits/merges are way off on that branch.
It would be great if you could somehow cooperate the changes.

Thank you!

**Comment by andreipaduroiu** [comment]
@RaulGracia @jkhalack @kevinhan88 
What is the status of this PR? I see no activity in 3 months. Last time I asked @RaulGracia said we still need to reach consensus. 

Please update this PR if it's still relevant or close it if we are not planning to anything. It has accumulated a lot of conflicts since we've done quite a bit of work in the Metrics space since that time.

**Comment by RaulGracia** [comment]
You are right @andreipaduroiu, I'm not aware of any agreement on the best way of reporting `segment_count` metric and actually @kevinhan88 is responsible for this PR (from a while ago). I updated him with the info of this PR, so let's see if he has done further progress. But keep in mind that @jkhalack needs this PR merged so Pravega can report metrics correctly.

**Comment by andreipaduroiu** [comment]
If @jkhalack needs this merged then why has this been opened for 6 months? @fpj what is the priority of this? 0.5?

**Comment by andreipaduroiu** [comment]
Ping! I will close this on Sept 9 if I don't hear anything regarding whether we want to proceed with this.

**Comment by kevinhan88** [comment]
Yes, agree this PR should be closed for the following reasons:
1. https://github.com/pravega/pravega/pull/3995 has partially solved the problem.
2. It's been proven this approach is not practical - controller report active segments - not only the operation to aggregate active segments is too expensive for a metric,  but also the synchronization of the aggregation among multiple controller is too expensive for a metric. This is the main reason why this PR got stuck for such a long time, and it's unlikely this PR will be merged in future.
3. I've talked with Andrei briefly and also mentioned it to the team a couple of times - instead we can let segmentstore to report the number of segments being read/written, and let the metrics dashboard to aggregate the total number of CURRENTLY being accessed segments. This dynamic number plus the real 'activeSegments' retrieved from API will give the whole picture to operation.

We can close this PR but leave https://github.com/pravega/pravega/issues/1761 open. I'll bring up another round of discussion for the approach above. If an agreement can be reached, then a new PR will be created. Regardless this PR can be closed.
